cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo

```diff
@@ -0,0 +1,187 @@
+    }
+
+    // ================== CONSTRUCTOR ================================
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState, summary_stats_address: ContractAddress, feed_type_id: FeedTypeId,
+    ) {
+        self.initializer(summary_stats_address, feed_type_id);
+    }
+    // ================== PUBLIC ABI ==================
+
+    #[abi(embed_v0)]
+    pub impl TwapRouterImpl of IFeedTypeRouter<ContractState> {
+        /// Returns the feed type id of the current router.
+        fn get_feed_type_id(self: @ContractState) -> FeedTypeId {
+            self.feed_type.read().id()
+        }
+
+        /// Returns the update for the feed as bytes.
+        fn get_data(self: @ContractState, feed: Feed) -> Bytes {
+            let pair_id = feed.pair_id;
+            let data_type = match self.simple_data_type.read() {
+                SimpleDataType::Spot => DataType::SpotEntry(pair_id),
+                SimpleDataType::Perp => DataType::FutureEntry((pair_id, 0))
+            };
+            let aggregation_mode = self.aggregation_mode.read();
+            let duration = self.duration.read().as_seconds();
+            let start_timestamp = get_block_timestamp() - duration;
+
+            let (twap_price, decimals) = self
+                .call_calculate_twap(data_type, aggregation_mode, start_timestamp, duration);
+
+            let mut update = BytesTrait::new_empty();
+            update.append_u256(feed.id().into());
+            update.append_u64(get_block_timestamp());
+            update.append_u16(0); // TODO: num sources ?
+            update.append_u8(decimals.try_into().unwrap());
+            update.append_u256(twap_price.into());
+            update.append_u256(0); // TODO: time period ?
+            update.append_u256(0); // TODO: start price ?
+            update.append_u256(0); // TODO: end price ?
+            update.append_u256(0); // TODO: total volume ?
+            update.append_u256(0); // TODO: number of data points ?
+
+            update
+        }
+    }
+
+
+    // ================== PRIVATE IMPLEMENTATIONS ==================
+
+    #[generate_trait]
+    impl InternalImpl of InternalTrait {
+        /// Initializes the contract storage.
+        /// Called only once by the constructor.
+        fn initializer(
+            ref self: ContractState,
+            summary_stats_address: ContractAddress,
+            feed_type_id: FeedTypeId,
+        ) {
+            // [Check] Contracts are not zero
+            assert(!summary_stats_address.is_zero(), errors::SUMMARY_STATS_IS_ZERO);
+
+            // [Check] Feed type id is valid
+            let feed_type = match FeedTypeTrait::from_id(feed_type_id) {
+                Result::Ok(f) => f,
+                Result::Err(e) => panic_with_felt252(e.into())
+            };
+            // [Check] Feed type variant & extract parameters
+            let (simple_data_type, aggregation_mode, duration) = match feed_type {
```

Comment:
This is a bit too complicated, I'd just have imagined that the parameters of the twap are given in the constructor and then used in the get_data function. No need to match on the variant or whatever

Reply:
Mmmmh, I agree but I'm not sure either:

1. We can indeed pass arbitrary parameters with the constructor but we would loose all traceability for which routers ***can*** exist,
2. By using only the `feed_type_id`, we're guaranteed to be able to use this id in the `AssetClassRouter` and find a unique contract.
3. It also mean that by looking at a documentation (or the code, as a developer), you can find all the possible `feed_type_id`  and it's a 1 <=> 1 matching with the routers that are (or can be) deployed, there can't be more definitions "hidden" or that we have "to guess" because the constructor allows anything.

But I agree that it makes the process of creating a new feed type more challenging (add the new variant in the types + the arm match in the router contract). I don't think we will create tons of them though.

any thoughts @JordyRo1 ?

Reply:
I do agree with @akhercha on this point. Indeed, it makes the creation process more complicated, but a way better tracability, since we already know the `feed_type_id` used in the entire process. But since we will be the only allowed to add to the route/deploy contract for now, it's just personal preference, since in both case we will have to write the feed -> feed_id_type conversion function, either before the initialization of the new contract (@akhercha) or before adding to the associated router (@EvolveArt )